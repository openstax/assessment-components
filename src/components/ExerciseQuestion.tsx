import { AvailablePoints, ID, Question as QuestionType, Task } from "../types";
import Button from "./Button";
import { Content } from "./Content";
import { Question } from './Question';
import { StepCardFooter } from "./StepCardFooter";

export interface ExerciseQuestionProps {
  task?: Task;
  question: QuestionType;
  answer_id_order?: ID[]; // @TODO harmonize these props with Exercise
  questionNumber: number;
  choicesEnabled: boolean;
  hasMultipleAttempts: boolean;
  onAnswerChange: () => void;
  onAnswerSave: () => void;
  onNextStep: () => void;
  feedback_html: string;
  correct_answer_feedback_html: string;
  is_completed: boolean;
  correct_answer_id: ID;
  incorrectAnswerId: ID;
  answerId?: ID;
  attempts_remaining: number;
  published_comments?: string;
  detailedSolution?: string;
  canAnswer: boolean;
  needsSaved: boolean;
  canUpdateCurrentStep: boolean;
  attempt_number: number;
  apiIsPending: boolean;
  displaySolution: boolean;
  available_points: AvailablePoints;
  exercise_uid: string;
  free_response?: string;
}

const AttemptsRemaining = ({ count }: { count: number }) => {
  return (
    <div>{count} attempt{count === 1 ? '' : 's'} left</div>
  );
}

const PublishedComments = ({ published_comments }: { published_comments?: string }) => {
  if (!published_comments) { return null; }

  return (
    <div>
      <strong>Feedback:</strong> {published_comments}
    </div>
  );
}

const SaveButton = (props: {
  disabled: boolean, isWaiting: boolean, attempt_number: number
} & React.ComponentPropsWithoutRef<'button'>) => (
  <Button
    {...props}
    waitingText="Savingâ€¦"
    isWaiting={props.isWaiting}
    data-test-id="submit-answer-btn"
  >
    {props.attempt_number == 0 ? 'Submit' : 'Re-submit'}
  </Button>
);

const NextButton = (props: {
  canUpdateCurrentStep: boolean,
  onNextStep: ExerciseQuestionProps['onNextStep']
}) => {
  return (
    <Button onClick={props.onNextStep} data-test-id="continue-btn">
      {props.canUpdateCurrentStep ? 'Continue' : 'Next'}
    </Button>
  );
}

const FreeResponseReview = ({ free_response }: Pick<ExerciseQuestionProps, "free_response">) => {
  if (!free_response) { return null; }
  return (
    <>
      <div className="free-response">{free_response}</div>
    </>
  );
}

export const ExerciseQuestion = (props: ExerciseQuestionProps) => {
  const {
    question, task, answer_id_order, onAnswerChange, feedback_html, correct_answer_feedback_html,
    is_completed, correct_answer_id, incorrectAnswerId, choicesEnabled, questionNumber,
    answerId, hasMultipleAttempts, attempts_remaining, published_comments, detailedSolution,
    canAnswer, needsSaved, attempt_number, apiIsPending, onAnswerSave, onNextStep, canUpdateCurrentStep,
    displaySolution, available_points, free_response
  } = props;

  return (
    <div data-test-id="student-exercise-question">
      <Question
        task={task}
        question={question}
        answerIdOrder={answer_id_order}
        choicesEnabled={choicesEnabled}
        answer_id={answerId}
        questionNumber={questionNumber}
        onChange={onAnswerChange}
        feedback_html={feedback_html}
        correct_answer_feedback_html={correct_answer_feedback_html}
        correct_answer_id={is_completed ? correct_answer_id : null}
        incorrectAnswerId={incorrectAnswerId}
        className="step-card-body"
        hideAnswers={false}
        displayFormats={false}
        displaySolution={displaySolution}
      >
        <FreeResponseReview free_response={free_response} />
      </Question>
      <StepCardFooter className="step-card-footer">
        <div className="points">
          <strong>Points: {available_points}</strong>
          <span className="attempts-left">
            {hasMultipleAttempts &&
              attempts_remaining > 0 &&
              <AttemptsRemaining count={attempts_remaining} />}
          </span>
          <PublishedComments published_comments={published_comments} />
          {detailedSolution && (<div><strong>Detailed solution:</strong> <Content html={detailedSolution} /></div>)}
        </div>
        <div className="controls">
          {canAnswer && needsSaved ?
            <SaveButton
              disabled={apiIsPending || !answerId}
              isWaiting={apiIsPending}
              attempt_number={attempt_number}
              onClick={onAnswerSave}
            /> :
            <NextButton onNextStep={onNextStep} canUpdateCurrentStep={canUpdateCurrentStep} />}
        </div>
      </StepCardFooter>
    </div>
  );
}
