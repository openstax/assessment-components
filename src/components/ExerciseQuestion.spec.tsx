import { ExerciseQuestion, ExerciseQuestionProps } from './ExerciseQuestion';
import renderer from 'react-test-renderer';

describe('ExerciseQuestion', () => {
  let props: ExerciseQuestionProps;

  props = {
    question: {
      id: '1',
      stem_html: 'Is this a question?',
      collaborator_solutions: [],
      formats: [],
      stimulus_html: '',
      answers: [{
        id: '1',
        correctness: undefined,
        content_html: 'True',
      }, {
        id: '2',
        correctness: undefined,
        content_html: 'False',
      }],
      is_answer_order_important: false,
    },
    questionNumber: 1,
    choicesEnabled: false,
    hasMultipleAttempts: false,
    onAnswerChange: () => null,
    onAnswerSave: () => null,
    onNextStep: () => null,
    feedback_html: '',
    correct_answer_feedback_html: '',
    is_completed: false,
    correct_answer_id: '',
    incorrectAnswerId: '',
    multiPartGroup: [],
    answerId: '',
    attempts_remaining: 2,
    published_comments: '',
    detailedSolution: '',
    canAnswer: false,
    needsSaved: false,
    canUpdateCurrentStep: false,
    attempt_number: 0,
    apiIsPending: false,
    displaySolution: false,
    available_points: '',
    exercise_uid: '',
  }

  it('matches snapshot', () => {
    const tree = renderer.create(
      <ExerciseQuestion {...props} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
