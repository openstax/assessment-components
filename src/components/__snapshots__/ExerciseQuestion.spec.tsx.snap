// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseQuestion matches snapshot 1`] = `
<div
  data-test-id="student-exercise-question"
>
  <div
    className="sc-dkzDqf cedznK openstax-question step-card-body"
    data-question-number={1}
    data-test-id="question"
  >
    <div
      className="question-stem"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Is this a question?",
        }
      }
      data-question-number={1}
    />
    <div
      aria-label="Answer choices"
      className="answers-table"
      role="radiogroup"
    >
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer disabled"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={true}
            id="1-option-0"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-0"
          >
            <span
              aria-label="Choice A:"
              className="answer-letter-wrapper"
              data-answer-choice="A"
              data-test-id="answer-choice-A"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "True",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer disabled"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={true}
            id="1-option-1"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-1"
          >
            <span
              aria-label="Choice B:"
              className="answer-letter-wrapper"
              data-answer-choice="B"
              data-test-id="answer-choice-B"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "False",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM iGTyMe step-card-footer"
  >
    <div
      className="step-card-footer-inner"
    >
      <div
        className="points"
        role="status"
      >
        <strong>
          Points: 
          1.0
        </strong>
        <span
          className="attempts-left"
        />
        
      </div>
      <div
        className="controls"
      >
        <button
          className="sc-bczRLJ fLTvYy"
          data-test-id="continue-btn"
          onClick={[Function]}
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExerciseQuestion renders Re-submit button 1`] = `
<div
  data-test-id="student-exercise-question"
>
  <div
    className="sc-dkzDqf cedznK openstax-question step-card-body has-incorrect-answer"
    data-question-number={1}
    data-test-id="question"
  >
    <div
      className="question-stem"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Is this a question?",
        }
      }
      data-question-number={1}
    />
    <div
      aria-label="Answer choices"
      className="answers-table"
      role="radiogroup"
    >
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer answer-selected"
        >
          <input
            checked={true}
            className="answer-input-box"
            disabled={false}
            id="1-option-0"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-0"
          >
            <span
              aria-label="Selected Choice A:"
              className="answer-letter-wrapper"
              data-answer-choice="A"
              data-test-id="answer-choice-A"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "True",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer answer-incorrect"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={false}
            id="1-option-1"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-1"
          >
            <span
              aria-label="Choice B:"
              className="answer-letter-wrapper"
              data-answer-choice="B"
              data-test-id="answer-choice-B"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <div
                className="sc-gsnTZi jsjSLp"
              >
                <span>
                  Incorrect Answer
                </span>
              </div>
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "False",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM iGTyMe step-card-footer"
  >
    <div
      className="step-card-footer-inner"
    >
      <div
        className="points"
        role="status"
      >
        <strong>
          Points: 
          1.0
        </strong>
        <span
          className="attempts-left"
        />
        
      </div>
      <div
        className="controls"
      >
        <button
          className="sc-bczRLJ fLTvYy"
          data-test-id="submit-answer-btn"
          disabled={false}
          onClick={[Function]}
        >
          Re-submit
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExerciseQuestion renders Save button 1`] = `
<div
  data-test-id="student-exercise-question"
>
  <div
    className="sc-dkzDqf cedznK openstax-question step-card-body has-incorrect-answer"
    data-question-number={1}
    data-test-id="question"
  >
    <div
      className="question-stem"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Is this a question?",
        }
      }
      data-question-number={1}
    />
    <div
      aria-label="Answer choices"
      className="answers-table"
      role="radiogroup"
    >
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer answer-selected"
        >
          <input
            checked={true}
            className="answer-input-box"
            disabled={false}
            id="1-option-0"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-0"
          >
            <span
              aria-label="Selected Choice A:"
              className="answer-letter-wrapper"
              data-answer-choice="A"
              data-test-id="answer-choice-A"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "True",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer answer-incorrect"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={false}
            id="1-option-1"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-1"
          >
            <span
              aria-label="Choice B:"
              className="answer-letter-wrapper"
              data-answer-choice="B"
              data-test-id="answer-choice-B"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <div
                className="sc-gsnTZi jsjSLp"
              >
                <span>
                  Incorrect Answer
                </span>
              </div>
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "False",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM iGTyMe step-card-footer"
  >
    <div
      className="step-card-footer-inner"
    >
      <div
        className="points"
        role="status"
      >
        <strong>
          Points: 
          1.0
        </strong>
        <span
          className="attempts-left"
        />
        
      </div>
      <div
        className="controls"
      >
        <button
          className="sc-bczRLJ fLTvYy"
          data-test-id="submit-answer-btn"
          disabled={false}
          onClick={[Function]}
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExerciseQuestion renders Submit & continue button 1`] = `
<div
  data-test-id="student-exercise-question"
>
  <div
    className="sc-dkzDqf cedznK openstax-question step-card-body has-incorrect-answer"
    data-question-number={1}
    data-test-id="question"
  >
    <div
      className="question-stem"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Is this a question?",
        }
      }
      data-question-number={1}
    />
    <div
      aria-label="Answer choices"
      className="answers-table"
      role="radiogroup"
    >
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer answer-selected"
        >
          <input
            checked={true}
            className="answer-input-box"
            disabled={false}
            id="1-option-0"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-0"
          >
            <span
              aria-label="Selected Choice A:"
              className="answer-letter-wrapper"
              data-answer-choice="A"
              data-test-id="answer-choice-A"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "True",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer answer-incorrect"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={false}
            id="1-option-1"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-1"
          >
            <span
              aria-label="Choice B:"
              className="answer-letter-wrapper"
              data-answer-choice="B"
              data-test-id="answer-choice-B"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <div
                className="sc-gsnTZi jsjSLp"
              >
                <span>
                  Incorrect Answer
                </span>
              </div>
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "False",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM iGTyMe step-card-footer"
  >
    <div
      className="step-card-footer-inner"
    >
      <div
        className="points"
        role="status"
      >
        <strong>
          Points: 
          1.0
        </strong>
        <span
          className="attempts-left"
        />
        
      </div>
      <div
        className="controls"
      >
        <button
          className="sc-bczRLJ fLTvYy"
          data-test-id="submit-answer-btn"
          disabled={false}
          onClick={[Function]}
        >
          Submit & continue
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExerciseQuestion renders all attempts remaining 1`] = `
<div
  data-test-id="student-exercise-question"
>
  <div
    className="sc-dkzDqf cedznK openstax-question step-card-body"
    data-question-number={1}
    data-test-id="question"
  >
    <div
      className="question-stem"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Is this a question?",
        }
      }
      data-question-number={1}
    />
    <div
      aria-label="Answer choices"
      className="answers-table"
      role="radiogroup"
    >
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={false}
            id="1-option-0"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-0"
          >
            <span
              aria-label="Choice A:"
              className="answer-letter-wrapper"
              data-answer-choice="A"
              data-test-id="answer-choice-A"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "True",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={false}
            id="1-option-1"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-1"
          >
            <span
              aria-label="Choice B:"
              className="answer-letter-wrapper"
              data-answer-choice="B"
              data-test-id="answer-choice-B"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "False",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM iGTyMe step-card-footer"
  >
    <div
      className="step-card-footer-inner"
    >
      <div
        className="points"
        role="status"
      >
        <strong>
          Points: 
          1.0
        </strong>
        <span
          className="attempts-left"
        >
          <div>
            2
             attempt
            s
             left
          </div>
        </span>
        
      </div>
      <div
        className="controls"
      >
        <button
          className="sc-bczRLJ fLTvYy"
          data-test-id="continue-btn"
          onClick={[Function]}
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExerciseQuestion renders continue button (unused?) 1`] = `
<div
  data-test-id="student-exercise-question"
>
  <div
    className="sc-dkzDqf cedznK openstax-question step-card-body has-incorrect-answer"
    data-question-number={1}
    data-test-id="question"
  >
    <div
      className="question-stem"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Is this a question?",
        }
      }
      data-question-number={1}
    />
    <div
      aria-label="Answer choices"
      className="answers-table"
      role="radiogroup"
    >
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer disabled"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={true}
            id="1-option-0"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-0"
          >
            <span
              aria-label="Choice A:"
              className="answer-letter-wrapper"
              data-answer-choice="A"
              data-test-id="answer-choice-A"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "True",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer disabled answer-incorrect"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={true}
            id="1-option-1"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-1"
          >
            <span
              aria-label="Choice B:"
              className="answer-letter-wrapper"
              data-answer-choice="B"
              data-test-id="answer-choice-B"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <div
                className="sc-gsnTZi jsjSLp"
              >
                <span>
                  Incorrect Answer
                </span>
              </div>
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "False",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM iGTyMe step-card-footer"
  >
    <div
      className="step-card-footer-inner"
    >
      <div
        className="points"
        role="status"
      >
        <strong>
          Points: 
          1.0
        </strong>
        <span
          className="attempts-left"
        />
        
      </div>
      <div
        className="controls"
      >
        <button
          className="sc-bczRLJ fLTvYy"
          data-test-id="continue-btn"
          onClick={[Function]}
        >
          Continue
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExerciseQuestion renders detailed solution and published comments 1`] = `
<div
  data-test-id="student-exercise-question"
>
  <div
    className="sc-dkzDqf cedznK openstax-question step-card-body has-correct-answer has-incorrect-answer"
    data-question-number={1}
    data-test-id="question"
  >
    <div
      className="question-stem"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Is this a question?",
        }
      }
      data-question-number={1}
    />
    <div
      aria-label="Answer choices"
      className="answers-table"
      role="radiogroup"
    >
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer disabled answer-correct"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={true}
            id="1-option-0"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-0"
          >
            <span
              aria-label="Choice A:"
              className="answer-letter-wrapper"
              data-answer-choice="A"
              data-test-id="answer-choice-A"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <div
                className="sc-gsnTZi evWjfC"
              >
                <span>
                  Correct Answer
                </span>
              </div>
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "True",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer disabled answer-incorrect"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={true}
            id="1-option-1"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-1"
          >
            <span
              aria-label="Choice B:"
              className="answer-letter-wrapper"
              data-answer-choice="B"
              data-test-id="answer-choice-B"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <div
                className="sc-gsnTZi jsjSLp"
              >
                <span>
                  Incorrect Answer
                </span>
              </div>
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "False",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM iGTyMe step-card-footer"
  >
    <div
      className="step-card-footer-inner"
    >
      <div
        className="points"
        role="status"
      >
        <strong>
          Points: 
          1.0
        </strong>
        <span
          className="attempts-left"
        />
        <div>
          <strong>
            Feedback:
          </strong>
           
          Teacher feedback
        </div>
        <div>
          <strong>
            Detailed solution:
          </strong>
           
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "A detailed solution",
              }
            }
          />
        </div>
      </div>
      <div
        className="controls"
      >
        <button
          className="sc-bczRLJ fLTvYy"
          data-test-id="continue-btn"
          onClick={[Function]}
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExerciseQuestion renders free response 1`] = `
<div
  data-test-id="student-exercise-question"
>
  <div
    className="sc-dkzDqf cedznK openstax-question step-card-body"
    data-question-number={1}
    data-test-id="question"
  >
    <div
      className="question-stem"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Is this a question?",
        }
      }
      data-question-number={1}
    />
    <div
      className="free-response"
    >
      A free response
    </div>
    <div
      aria-label="Answer choices"
      className="answers-table"
      role="radiogroup"
    >
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer disabled"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={true}
            id="1-option-0"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-0"
          >
            <span
              aria-label="Choice A:"
              className="answer-letter-wrapper"
              data-answer-choice="A"
              data-test-id="answer-choice-A"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "True",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer disabled"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={true}
            id="1-option-1"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-1"
          >
            <span
              aria-label="Choice B:"
              className="answer-letter-wrapper"
              data-answer-choice="B"
              data-test-id="answer-choice-B"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "False",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM iGTyMe step-card-footer"
  >
    <div
      className="step-card-footer-inner"
    >
      <div
        className="points"
        role="status"
      >
        <strong>
          Points: 
          1.0
        </strong>
        <span
          className="attempts-left"
        />
        
      </div>
      <div
        className="controls"
      >
        <button
          className="sc-bczRLJ fLTvYy"
          data-test-id="continue-btn"
          onClick={[Function]}
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExerciseQuestion renders no attempts remaining 1`] = `
<div
  data-test-id="student-exercise-question"
>
  <div
    className="sc-dkzDqf cedznK openstax-question step-card-body has-incorrect-answer"
    data-question-number={1}
    data-test-id="question"
  >
    <div
      className="question-stem"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Is this a question?",
        }
      }
      data-question-number={1}
    />
    <div
      aria-label="Answer choices"
      className="answers-table"
      role="radiogroup"
    >
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer disabled"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={true}
            id="1-option-0"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-0"
          >
            <span
              aria-label="Choice A:"
              className="answer-letter-wrapper"
              data-answer-choice="A"
              data-test-id="answer-choice-A"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "True",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer disabled answer-incorrect"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={true}
            id="1-option-1"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-1"
          >
            <span
              aria-label="Choice B:"
              className="answer-letter-wrapper"
              data-answer-choice="B"
              data-test-id="answer-choice-B"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <div
                className="sc-gsnTZi jsjSLp"
              >
                <span>
                  Incorrect Answer
                </span>
              </div>
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "False",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM iGTyMe step-card-footer"
  >
    <div
      className="step-card-footer-inner"
    >
      <div
        className="points"
        role="status"
      >
        <strong>
          Points: 
          1.0
        </strong>
        <span
          className="attempts-left"
        />
        
      </div>
      <div
        className="controls"
      >
        <button
          className="sc-bczRLJ fLTvYy"
          data-test-id="continue-btn"
          onClick={[Function]}
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ExerciseQuestion renders some attempts remaining 1`] = `
<div
  data-test-id="student-exercise-question"
>
  <div
    className="sc-dkzDqf cedznK openstax-question step-card-body has-incorrect-answer"
    data-question-number={1}
    data-test-id="question"
  >
    <div
      className="question-stem"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Is this a question?",
        }
      }
      data-question-number={1}
    />
    <div
      aria-label="Answer choices"
      className="answers-table"
      role="radiogroup"
    >
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={false}
            id="1-option-0"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-0"
          >
            <span
              aria-label="Choice A:"
              className="answer-letter-wrapper"
              data-answer-choice="A"
              data-test-id="answer-choice-A"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "True",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
      <div
        className="openstax-answer"
      >
        <section
          className="answers-answer answer-incorrect"
        >
          <input
            checked={false}
            className="answer-input-box"
            disabled={false}
            id="1-option-1"
            name="1-options"
            onChange={[Function]}
            type="radio"
          />
          <label
            className="answer-label"
            htmlFor="1-option-1"
          >
            <span
              aria-label="Choice B:"
              className="answer-letter-wrapper"
              data-answer-choice="B"
              data-test-id="answer-choice-B"
            />
            <div
              aria-atomic="true"
              aria-live="polite"
              className="answer-answer"
              role="status"
            >
              <div
                className="sc-gsnTZi jsjSLp"
              >
                <span>
                  Incorrect Answer
                </span>
              </div>
              <span
                className="answer-content"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "False",
                  }
                }
              />
            </div>
          </label>
        </section>
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM iGTyMe step-card-footer"
  >
    <div
      className="step-card-footer-inner"
    >
      <div
        className="points"
        role="status"
      >
        <strong>
          Points: 
          1.0
        </strong>
        <span
          className="attempts-left"
        >
          <div>
            1
             attempt
            
             left
          </div>
        </span>
        
      </div>
      <div
        className="controls"
      >
        <button
          className="sc-bczRLJ fLTvYy"
          data-test-id="continue-btn"
          onClick={[Function]}
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;
